dali/pipeline/operator/builtin/external_source_test.cc

```diff
@@ -22,9 +24,189 @@
 #include "dali/test/dali_test_config.h"
+#include "dali/c_api.h"
 
 namespace dali {
 
+namespace {
+
+/**
+ * Load the test file as a binary string.
+ */
+std::string LoadTestFile(const std::string &test_file_path) {
+  std::ifstream fin(test_file_path, std::ios::binary);
+  if (!fin)
+    throw std::runtime_error(std::string("Failed to open file: ") + test_file_path);
+  std::stringstream ss;
+  ss << fin.rdbuf();
+  return ss.str();
+}
+
+}  // namespace
+
+template<typename Backend>
+class ExternalSourceBasicTest : public ::testing::Test {
+ protected:
+  static constexpr bool is_cpu = std::is_same_v<Backend, CPUBackend>;
+
+  void CreateAndSerializePipeline(bool repeat_last) {
+    auto pipeline = std::make_unique<Pipeline>(batch_size_, num_threads_, device_id_, false, 1,
+                                               false);
+    pipeline->AddOperator(
+            OpSpec("ExternalSource")
+                    .AddArg("repeat_last", repeat_last)
+                    .AddArg("device", std::string{is_cpu ? "cpu" : "gpu"})
+                    .AddArg("name", input_name_)
+                    .AddOutput(input_name_, std::string{is_cpu ? "cpu" : "gpu"}),
+            input_name_);
+
+    std::vector<std::pair<std::string, std::string>> outputs = {
+            {input_name_, is_cpu ? "cpu" : "gpu"},
+    };
+
+    pipeline->SetOutputDescs(outputs);
+
+    serialized_pipeline_ = pipeline->SerializeToProtobuf();
+  }
+
+
+  void FeedExternalInput(daliPipelineHandle *h) {
+    daliSetExternalInputBatchSize(h, input_name_.c_str(), batch_size_);
+    std::vector<int64_t> shapes;
+    std::string data;
+    for (int i = 0; i < batch_size_; i++) {
+      auto sample = LoadTestFile(test_files_[i % test_files_.size()]);
+      shapes.push_back(static_cast<int64_t>(sample.length()));
+      data += sample;
+    }
+    if constexpr (is_cpu) {
+      daliSetExternalInput(h, input_name_.c_str(), device_type_t::CPU, data.data(),
+                           dali_data_type_t::DALI_UINT8, shapes.data(), 1, nullptr,
+                           DALI_ext_force_copy);
+    } else {
+      thrust::host_vector<uint8_t> data_cpu{data.begin(), data.end()};
+      thrust::device_vector<uint8_t> data_gpu = data_cpu;
+      daliSetExternalInput(h, input_name_.c_str(), device_type_t::GPU,
+                           thrust::raw_pointer_cast(data_gpu.data()),
+                           dali_data_type_t::DALI_UINT8, shapes.data(), 1, nullptr,
+                           DALI_ext_force_copy);
+    }
+  }
+
+
+  /**
+   * Check, if the "depleted" trace exists.
+   */
+  void DoesDepletedTraceExist(daliPipelineHandle *h) {
+    // The "depleted" trace should always exist.
+    ASSERT_TRUE(daliHasOperatorTrace(h, input_name_.c_str(), depleted_trace_name_.c_str()));
+  }
+
+
+  /**
+   * Verify the value of "depleted" trace.
+   * @param shall_be_depleted Expected value.
+   */
+  void IsDepletedTraceCorrect(daliPipelineHandle *h, bool shall_be_depleted) {
```

Comment:
```suggestion
  void CheckDepletedTraceCorrect(daliPipelineHandle *h, bool shall_be_depleted) {
```
would be a better name

Reply:
Done